### PR TITLE
Unify AI resourcemap analyzer(s)

### PR DESCRIPTION
**Pull request by sprunk**
The skirmish AI interface exposes resource analyzers to find resource spots.
Each resource type has its own analyzer instance. But only metal has a map
with distribution. Energy doesn't (geo spots aren't built into the map, but
are actually features), and hypothetical future extra resource types also
wouldn't.

Future commits will make it so that other resource types can be extracted,
but will still use the "metalmap", which will become a generic resource map.

This commit removes the energymap analyser and makes it so that asking about
energy (and other future resources) returns generic resource distribution,
i.e. the metal map.

It maintains the logic around object lifetime etc very closely (hence the
std::optional) because I am scared of things like save/load interaction.
Anybody is free to improve that part.

This is a logic change but I don't expect any AI to actually ask about the
energy map (because why would they, it's empty). Circuit at least doesn't.
So there should be no observed change in practice.

### CODE REVIEW

**Comment by saurtron** [comment]
> But only metal has a map
> with distribution. Energy doesn't (geo spots aren't built into the map, but
> are actually features), and hypothetical future extra resource types also
> wouldn't.


Why would other resource types not have a specific map? Or is your plan to make the metal map actually hold different resource distributions in the future?

**Comment by sprunk** [comment]
>  Or is your plan to make the metal map actually hold different resource distributions in the future?

tl;dr yes.

Currently games can make their units do `extractsMetal = 1.23` and the unit extracts the "metal" resource. I plan to change it to `extracts = { moonrocks = 1.23, blood = 4.56, requisition = 7.89 }` so that games can make units extract any resource. Maps will stay as-is, and `extractsMetal = x` will be equivalent to `extracts = { metal = x }` for backwards compatibility.

In theory this leaves you with still just a single distribution map, which has some problems:
* there is only one distribution map, so you can at most say "there are some resources here" but you can put either a crystal mine or a blood pump and they will extract all the same. How do you get separate distributions?
* what if you put something with `extracts = { copper = 10 }` and afterwards another unit with `extracts = { coal = 13, copper = 4 }` in the same spot, do they stack and how if so? Does the second unit extract 3 coal 0 copper, or 13 coal 0 copper, or something else yet? Importantly, this is a game design question with multiple reasonable answers so games should be answering it via Lua.

So there's an intuition that perhaps letting maps specify multiple distribution maps would be an improvement. But:
* after metal and energy are unhardcoded from engine, it is games who will be the ones to define what resource types exist, not maps. Say there's a map with 2 distributions, one for "metal" assuming it is a common resource, and one for "gems" assuming it is rare. But then somebody makes a game with ancient romans which has "wood" as a common resource, "stone" as a sort of mid tier, and "metal" as a super rare. How does this work out, what gets mapped to what?
* if all resource distributions are equal (or if a single place just has overlapping resources) then it's a situation equivalent to the proposed single generic resource situation, which makes the "how do they stack" question remain valid. 
* it would require a change in map format, which cascades into more work, e.g. you'd need map compilers to support it.
* speaking of compilation, it's annoying to recompile a map. Wouldn't it be great if whatever new format we come up with involved just cramming the extra bitmaps into the VFS?
* and then what this all boils down to in practice: most maps would still just have 1 distribution map for whatever is the "common" resource anyway, and games would mostly have 1 territorially-distributed resource; and then it usually turns out to be just a list of discrete resource spots that can be a list of XZ coordinates in a lua text file. Good support to relatively easily add the extras should definitely exist but the engine doesn't have to go a bunch of extra miles to add native systems for that purpose. Can we let Lua specify distribution maps?

So my plan to support extra distribution maps is basically that we get it for free with #1561:
* is an absurdly powerful generic feature with many use cases, resource distribution is just one of them.
* lets gamedevs and mappers do whatever mapping is appropriate. Mappers could come up with a standard, e.g. maps would put something like `./mapconfig/resources/rare_secondary.png` and games would be able to parse it into whichever resource they know best to be secondary and rare. Or the reverse, games could come up with a standard ("put skulls distribution in `./mapconfig/sledgehammer_50K/skulls_for_the_skull_throne.png`") and mappers could conform.
* is just a file on the VFS so working with it doesn't require recompilation, just a `/luafoo reload`.
* is just a file on the VFS so it is trivial to parse (e.g. for website map info), your native os file manager can even display it.
* is just a file on the VFS so games can override locally without having to re-release a map.
